### PR TITLE
Document function signatures in C ABI

### LINKED ISSUE #160: Document rules for complex params passed to / returned from functions in C ABI
We're trying to align representation of structures/unions/arrays when passed to functions between different toolchains, but such representation is currently not documented in `BasicCABI.md` and only implicitly exists as part of LLVM implementation.

### PR DESCRIPTION
**Pull request by tlively**
Fixes #160.

### CODE REVIEW

File: BasicCABI.md [hunk 1]
```diff
@@ -156,6 +156,15 @@ is used as the frame pointer or base pointer. This functionality is not needed f
 wasm VM must do this in any case); however it may still be desirable to allow this functionality for debugging
 or in-field crash reporting. Future ABIs may designate a convention for determining frame size and local usage.
 
+### Function signatures
+
+Scalar types passed as arguments are passed via WebAssembly function parameters and returned via WebAssembly function
+results of their corresponding Wasm value types. `long double`, which has no corresponding Wasm value type, is passed
+as a pair of `i64` parameters and returned indirectly via a pointer to memory prepended to the function parameters.
+Empty structs and unions are ignored in both arguments and return positions. Structs and unions with a single element
```
Comment (RReverser) [inline BasicCABI.md:164]: > Structs and unions with a single element are passed and returned the same way their element would be.

We should probably mention a case with explicit alignment override too. E.g.:

```
struct S {
    _Alignas(8)
    int x;
};
```

seems to be passed by pointer, despite having a single element.

File: BasicCABI.md [hunk 2]
```diff
@@ -156,6 +156,15 @@ is used as the frame pointer or base pointer. This functionality is not needed f
 wasm VM must do this in any case); however it may still be desirable to allow this functionality for debugging
 or in-field crash reporting. Future ABIs may designate a convention for determining frame size and local usage.
 
+### Function signatures
+
+Scalar types passed as arguments are passed via WebAssembly function parameters and returned via WebAssembly function
+results of their corresponding Wasm value types. `long double`, which has no corresponding Wasm value type, is passed
```
Comment (RReverser) [inline BasicCABI.md:162]: General nitpick: I'd format this as a list / somehow split up into lines for easier readability.

File: BasicCABI.md (hunk 1 above)
Comment (RReverser) [reply BasicCABI.md:164]: I'm guessing there are also some rules for contents of the struct, in particular when arrays are involved?

E.g.

```
struct S {
    int x[1];
};
```

seems to be passed by value (as just `int`) but

```
struct S {
    int x[2];
};
```

is passed as a pointer.

Comment (RReverser) [reply BasicCABI.md:164]: Not sure how to correctly put that into words. Something like "if struct recursively contains only a single scalar value"?

**Comment by tlively** [comment]
@RReverser I made it a table, ptal.

**Comment by sunfishcode** [comment]
Would it make sense to mention the upcoming multivalue ABI here, so that people building tools that know about C ABIs are aware that this change is coming?

File: BasicCABI.md [hunk 3]
```diff
@@ -156,6 +156,25 @@ is used as the frame pointer or base pointer. This functionality is not needed f
 wasm VM must do this in any case); however it may still be desirable to allow this functionality for debugging
 or in-field crash reporting. Future ABIs may designate a convention for determining frame size and local usage.
 
+### Function signatures
+
+Types can be passed directly via WebAssembly function parameters or indirectly
+via a pointer parameter that points to the value in memory. Similarly, types can
+either be returned directly from WebAssembly functions or returned indirectly
```
Comment (dschuff) [inline BasicCABI.md:163]: For byval IIRC LLVM currently guarantees that the callee will not modify the buffer. Do we want to make a similar guarantee in the ABI? For sret, it's probably worth making explicit that the caller must allocate the return buffer (even though it's sort of implied since the caller passes a pointer).

**Comment by RReverser** [comment]
> the upcoming multivalue ABI here, so that people building tools that know about C ABIs are aware that this change is coming?

I suspect multivalue ABI will be an additional ABI, rather than a replacement for the current one? (we'll still need this one for compatibility with older browsers / engines)

File: BasicCABI.md [hunk 4]
```diff
@@ -156,6 +156,25 @@ is used as the frame pointer or base pointer. This functionality is not needed f
...
+Type                         | Parameter     | Result   |
+-----------------------------|---------------|----------|
+scalar[1]                    | direct        | direct   |
+empty struct or union        | ignored       | ignored  |
+singleton struct or union[2] | direct        | direct   |
+other struct or union        | indirect      | indirect |
+
+[1] `long long double` is passed directly as two `i64` values.
+
+[2] Any struct or union that recursively contains just a single scalar value and
```
Comment (RReverser) [inline BasicCABI.md:175]: I think somewhere here we want to spell out what happens to arrays & VLAs too, since we don't mention them anywhere else. Maybe something like:

```suggestion
[2] Any struct or union that recursively (including nested structs, unions and arrays) contains just a single scalar value and
```

File: BasicCABI.md (hunk 3 above)
Comment (RReverser) [reply BasicCABI.md:163]: > For byval IIRC LLVM currently guarantees that the callee will not modify the buffer

Hmm, are you sure? When I compile code like:

```c
struct A {
    int x;
    int y;
};

void foo(struct A a) {
    a.x = 42;
}
```

it produces:
```
foo:                                    # @foo
        local.get       0
        i32.const       42
        i32.store       0
        end_function
```

which seems to modify the caller's buffer.

(Godbolt link: https://clang.godbolt.org/z/crKarv)

**Comment by sunfishcode** [comment]
Multivalue [is standardized](https://github.com/WebAssembly/proposals/blob/master/finished-proposals.md) and [widely supported](https://webassembly.org/roadmap/) now, so I'm hoping multivalue will become the default, even if we also retain the pre-multivalue ABI for compatibility.

File: BasicCABI.md [hunk 5]
```diff
@@ -156,6 +156,25 @@ is used as the frame pointer or base pointer. This functionality is not needed f
...
+either be returned directly from WebAssembly functions or returned indirectly
+via a pointer parameter prepended to the parameter list.
+
+Type                         | Parameter     | Result   |
+-----------------------------|---------------|----------|
+scalar[1]                    | direct        | direct   |
+empty struct or union        | ignored       | ignored  |
+singleton struct or union[2] | direct        | direct   |
+other struct or union        | indirect      | indirect |
+
```
Comment (RReverser) [inline BasicCABI.md:172]: Also just to mention explicitly:

```suggestion
other struct or union        | indirect      | indirect |
array                        | indirect      | N/A      |

```

**Comment by RReverser** [comment]
> 
> 
> Multivalue [is standardized](https://github.com/WebAssembly/proposals/blob/master/finished-proposals.md?rgh-link-date=2021-01-14T00%3A42%3A23Z) and [widely supported](https://webassembly.org/roadmap/) now, so I'm hoping multivalue will become the default, even if we also retain the pre-multivalue ABI for compatibility.

Right, that's my point - we'll still retain this ABI for compatibility.

Given that it already has "As mentioned in README, it's not the only possible C ABI. It is the ABI that the clang/LLVM WebAssembly backend is currently using, [...]" in the disclaimer, that seems sufficient and we can reword once we create a doc for multivalue C ABI.

File: BasicCABI.md (hunk 3 above)
Comment (dschuff) [reply BasicCABI.md:163]: Oh you are right; the (langref)[https://llvm.org/docs/LangRef.html#parameter-attributes] says that a hidden copy is created "between the caller and the callee" so that the caller's copy can't be modified but doesn't say how.
Our (implementation)[https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp#L902] creates the copy in the caller's frame, which is simple and makes sense.
So the machine ABI should probably specify that the caller may modify the copy that is passed to it, and therefore the callee must create a copy (or otherwise ensure that any modification doesn't affect the calling code, "as-if" there were a copy; maybe that part goes without saying?)

**Comment by dschuff** [comment]
This section should probably also mention varargs. They work similarly (https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp#L950) to byval: the caller assigns each argument an offset according to its alignment and size, creates a buffer in its stack frame (at compile time), and then at runtime copies each argument into its place and passes a pointer to the buffer (as the last argument).

**Comment by RReverser** [comment]
@dschuff Good idea, yeah. So, if I understand correctly, varargs are laid out as if they were fields in an extra struct passed to the function?

File: BasicCABI.md (hunk 3 above)
Comment (RReverser) [reply BasicCABI.md:163]: > that the caller may modify the copy that is passed to it, and therefore the callee must create a copy

I think you mean the other way around (callee...caller)?

Is this part of the ABI too? I thought it was just general semantics of the C language, rather than a machine-specific thing.

Comment (dschuff) [reply BasicCABI.md:163]: err... yes, the other way around. (callee may modify, so the caller creates a copy).

It is the semantics of the C language, but how it happens could vary between machines. It could just as easily be that the caller just passes a pointer to its own copy, and then the ABI requires that the callee not modify that copy (forcing it to make its own copy if necessary). Either way ABI means that the caller's copy is not modified.

File: BasicCABI.md (hunk 5 above)
Comment (tlively) [reply BasicCABI.md:172]: I don't think there is any ambiguity about this in the C semantics, but adding this keeps the reader from having to think about that 👍

File: BasicCABI.md (hunk 3 above)
Comment (RReverser) [reply BasicCABI.md]: Interesting, makes sense.

File: BasicCABI.md [hunk 6]
```diff
@@ -159,21 +159,35 @@ or in-field crash reporting. Future ABIs may designate a convention for determin
...
+[2] Any struct or union that recursively (including through nested structs,
+unions, and arrays) contains just a single scalar value and is not specified to
+have greater than natural alignment.
+
+[3] This calling convention was defined before
+[multivalue](https://github.com/WebAssembly/multi-value) was standardized. A new
+default calling convention that changes this behavior and takes advantage of
+multivalue may be introduced in the future.
+
+Varargs are placed in a buffer by the caller and the last parameter to the
```
Comment (RReverser) [inline BasicCABI.md:187]: Is the memory layout of varargs machine-specific? (at least that's my understanding from Derek's link above)

If it is, probably worth documenting its layout here too.

Comment (dschuff) [reply BasicCABI.md:187]: by "machine-specific" you mean different on wasm vs other architectures? Yes, it's different between machines and OS/ABI on the same machine. Some ABIs make the varargs calling convention match the native calling convention (so that a varargs function can be called with a non-varargs prototype or vice versa). I think that's only useful if the calling convention also tolerates mismatches in number of arguments though (which ours doesn't) and even then it's doubtful.

Comment (dschuff) [reply BasicCABI.md:187]: As to whether the layout of this buffer is the same as if it were a struct, I'm not 100% sure about that, actually. It's certainly not implemented by the same code in the frontend that lays out structures. It might be worth looking into that a bit more before making a more detailed description.

Comment (RReverser) [reply BasicCABI.md:187]: > Yes, it's different between machines and OS/ABI on the same machine.

Yeah, that's what I meant. Since it it Wasm-specific, it's probably useful to document how varargs are laid out in that struct. As far as I can tell from the code, it's similar to alignment in a struct in that each vararg is simply aligned to its own alignment (like fields would be), but I can't tell from a quick glance whether varargs structure as a whole are also aligned to the most-aligned "field".

Comment (RReverser) [reply BasicCABI.md:187]: Looks like that code was written by @sunfishcode, maybe he can clarify the layout to document here?

**Comment by tlively** [comment]
How does that extra detail look, @RReverser?

**Comment by RReverser** [comment]
> How does that extra detail look, @RReverser?

LGTM. I, obviously, can't speak for its correctness / completeness, especially in light of me being confused by the `3 * sizeof(varargs)` allocation that Clang seems to be doing, but if you think this description is sufficiently correct, I'm happy to go with it.

**Review by RReverser — APPROVED** [review]

File: BasicCABI.md [hunk 7]
```diff
@@ -156,6 +156,41 @@ is used as the frame pointer or base pointer. This functionality is not needed f
...
+
+Type                         | Parameter     | Result   |
+-----------------------------|---------------|----------|
+scalar[1]                    | direct        | direct   |
+empty struct or union        | ignored       | ignored  |
+singleton struct or union[2] | direct        | direct   |
+other struct or union[3]     | indirect      | indirect |
+array                        | indirect      | N/A      |
+
+[1] `long long double` is passed directly as two `i64` values.
```
Comment (RReverser) [inline BasicCABI.md:176]:
```suggestion
[1] `long long double` and `__int128` are passed directly as two `i64` values.
```

Comment (RReverser) [reply BasicCABI.md:176]: I know it's compiler-specific extension, but seeing as it's supported by Clang and Rust C FFI, it's probably worth documenting.

**Comment by RReverser** [comment]
@tlively Thanks for documenting this!

File: BasicCABI.md [hunk 8]
```diff
@@ -158,12 +158,22 @@ or in-field crash reporting. Future ABIs may designate a convention for determin
...
+via a pointer parameter prepended to the parameter list.
+
+Type                         | Parameter     | Result   |
+-----------------------------|---------------|----------|
+scalar[1]                    | direct        | direct   |
+empty struct or union        | ignored       | ignored  |
+singleton struct or union[2] | direct        | direct   |
+other struct or union        | indirect      | indirect |
+
+[1] `long long double` is passed directly as two `i64` values.
```
Comment (bjorn3) [inline BasicCABI.md]: Why was this changed to refer to `long long double` rather than `long double`? `long long double` is not defined in the table of primitive types.

Comment (dschuff) [reply BasicCABI.md]: I think this should have been just `long double`. `long double` in LLVM is a 128-bit FP value, and the math for it is software-emulated since there is no fp128 scalar type in wasm. So it would make sense to pass as a pair of i64s (although I haven't double-checked). AFAIK there is no `long long double`.
